### DISCUSSION TITLE
TeaVM

### DISCUSSION THREAD

**Discussion by shannah** [webfx-project/webfx#21] [General]
I'm curious what the rationale was for choosing GWT over TeaVM.  Seems like GWT will always face the glass ceiling of not supporting threads or third-party jar dependencies (that haven't been built for GWT), since GWT works on the source code.  TeaVM supports threads, works on bytecode, which seems much more flexible.

**Comment by salmonb** [comment]
hi Steve, I made that choice at the start of the WebFX project back in 2015 or 2016, so the situation was quite different. According to the tests I conducted at that time, I got smaller and faster JS code with GWT than TeaVM, which was my main criteria. TeaVM was faster to compile, but this was not crucial for me, as JS compilation is not required that frequently for final users developing WebFX apps (OpenJFX can be used instead during the main development cycle). Also considering that Google was behind GWT, compared to a single developer behind TeaVM at that time, the balance went in favour of GWT in the end. I don't know if I made the right decision, it was just the prototyping phase, and I had no idea how far the project will go.

But I'm glad that TeaVM grew up since that time, and as you can see in our [roadmap](https://github.com/webfx-project/webfx/blob/main/.github/ROADMAP.md), we plan to support other transpilers including TeaVM. This is not our priority however for now, this may happen in a year or so. Is the TeaVM team interested in helping us providing that support?

**Comment by reportmill** [comment]
I'm a huge fan of TeaVM, but I would suggest CheerpJ as well. CheerpJ doesn't require the additional compilation step - developers can use the same jars for desktop and browser. And I could easily compile to WebFX in the browser with [SnapCode](https://reportmill.com/SnapCode).

I think CheerpJ is free for Open-source uses, though it's commercial for commercial uses. And they are limited to Java 8 for the next few months (hopefully they will support Java 17 by year end).

**Comment by ivan-velikanov** [comment]
@salmonb I noticed some teavm related committed you recently made. Overall looks like teavm support has increased a lot. In your project. In https://github.com/webfx-project/webfx?tab=readme-ov-file#status the Roadmap, it has mentioned teavm as just 20%, i think this number is quiet outdated. So broadly what is the actual status of teavm support?

**Reply by salmonb** [reply]
I'm investigating the teavm support indeed, and maybe I reached 40%.

Actually I first went to the wrong direction, which was to write a teavm module for each feature in [webfx-platform](https://github.com/webfx-project/webfx-platform) (you should see these modules at this time and this is probably the commits you're talking about).

However this is too much maintenance work on the long run, because this results in a code duplication (one code version for gwt/j2cl and another one for teavm). If I apply the same on the [webfx-kit](https://github.com/webfx-project/webfx/tree/main/webfx-kit) (the core of the JavaFX transpiler) which have a much more consequent codebase, I will end up maintaining 2 versions of the code with 2 different APIs (elemental2 for gwt/j2cl and JSO for TeaVM).

Now, I'm changing direction and try to write a polyfill of elemental2 for TeaVM - see the concept below:
<div align="center">
<img width="482" height="315" alt="image" src="https://github.com/user-attachments/assets/3acbed5e-43b7-4649-88d1-224c48e5df93" />
</div>
So it's just one piece of work that will open the door to the teavm support while keeping just one source of truth for the remaining codebase that will stay with elemental2.
It's a bit challenging because this requires some features that teavm doesn't have, such as mapping Java fields to JS fields.
But I'm currently working closely with Alexey (the author of teavm) to get this working.

There is still lots of work to complete this polyfill, but I think this is the right direction and maybe I will finish by the end of this year (considering I'm not full time on this task).

**Reply by ivan-velikanov** [reply]
This sounds very promising, especially the close collaboration sounds very encouraging for the overall ecosystem viability and long term benefits. Actually when I think about it, I think this might also reduce the size of the java wrapper shim layers for teavm as well. Because I guess elemental2 more easily/transparently maps to js, and also it might makes direct typescript type definitions to java api convertion (using custom generators) more easier in teavm. I might be wrong, but I think it will benefit both ecosystems and I don't see drawbacks as of now.

**Comment by salmonb** [comment]
Please see this [announcement](https://github.com/webfx-project/webfx/discussions/54) and [blog](https://blog.webfx.dev/2025/11/17/teavm/) about TeaVM support in WebFX.

**Reply by ivan-velikanov** [reply]
@salmonb Thank you for keeping us posted, and congratulations for achieving this major milestone. My best wishes for continued success.

**Comment by reportmill** [comment]
Very cool! You should post this on Reddit/java.

**Reply by salmonb** [reply]
Good idea, but I'm not posting much content on Reddit, so I can't post on r/Java with my account (not enough karma).
But I could on r/JavaFX : https://www.reddit.com/r/JavaFX/comments/1p178xg/webfx_now_supports_teavm_bringing_webassembly_and/
